File: go/leet_code/85.go
// Package leet_code provides solutions for LeetCode problems.
package leet_code

// maximalRectangle calculates the largest rectangle that can be filled with '1's in the given matrix.
// The function uses a dynamic programming approach where it calculates the maximum area of the rectangle
// at each cell of the matrix. The area is calculated based on the height and width (left and right boundaries)
// of the rectangle that can be formed at each cell.
//
// Parameters:
//  - matrix: A 2D byte array where '1's represent filled cells and '0's represent empty cells.
//
// Returns:
//  - The area of the largest rectangle that can be filled with '1's.
func maximalRectangle(matrix [][]byte) int {
	// m and n represent the number of rows and columns in the matrix respectively.
	m, n := len(matrix), len(matrix[0])

	// heights, left, and right are arrays that store the height of the rectangle and the left and right boundaries of the rectangle at each column.
	heights := make([]int, n)
	left := make([]int, n)
	right := make([]int, n)

	// Initialize the right boundaries to the number of columns in the matrix.
	for j := 0; j < n; j++ {
		right[j] = n
	}

	// max stores the maximum area of the rectangle found so far.
	max := 0

	// Iterate over each cell in the matrix.
	for i := 0; i < m; i++ {
		// leftest and rightest store the leftmost and rightmost boundaries of the rectangle at the current row.
		leftest, rightest := 0, n - 1

		// Iterate over each column in the current row.
		for j := 0; j < n; j++ {
			// If the current cell is filled ('1'), update the height and left boundary of the rectangle at the current column.
			if matrix[i][j] == '1' {
				heights[j]++
				if leftest > left[j] {
					left[j] = leftest
				}
			} else {
				// If the current cell is empty ('0'), reset the height and left boundary of the rectangle at the current column and update leftest.
				heights[j] = 0
				left[j] = 0
				leftest = j + 1
			}

			// Update the right boundary of the rectangle at the current column from the right side of the row.
			if matrix[i][n-1-j] == '1' {
				if rightest < right[n-1-j] {
					right[n-1-j] = rightest
				}
			} else {
				right[n-1-j] = n
				rightest = n - 1 - j - 1
			}
		}

		// Calculate the area of the rectangle at each column and update max if a larger area is found.
		for j := 0; j < n; j++ {
			area := (right[j] - left[j] + 1) * heights[j]
			if area > max {
				max = area
			}
		}
	}

	// Return the maximum area of the rectangle found.
	return max
}

File: java/leet_code/85.java
import java.util.Arrays;

/** This class provides a solution for finding the maximal rectangle in a binary matrix. */
class Solution {
    /**
     * Finds the maximal rectangle in a binary matrix.
     *
     * @param matrix The binary matrix.
     *
     * @return The area of the maximal rectangle.
     */
    public static int maximalRectangle(final char[][] matrix) {
        // If the matrix is null or empty, return 0
        if (null == matrix || 0 == matrix.length || 0 == matrix[0].length) {
            return 0;
        }

        final int m = matrix.length;
        final int n = matrix[0].length;

        // Initialize arrays to store heights and boundaries of rectangles
        final int[] heights = new int[n];
        final int[] leftBoundaries = new int[n];
        final int[] rightBoundaries = new int[n];
        Arrays.fill(rightBoundaries, n);

        int maxRectangle = 0;

        // Iterate over each row in the matrix
        for (final char[] chars : matrix) {
            final int left = 0;

            // Update heights and left boundaries for each row
            Solution.updateHeightsAndLeftBoundaries(chars, heights, leftBoundaries, left);

            // Update right boundaries for each row
            Solution.updateRightBoundaries(chars, rightBoundaries, n);

            // Calculate the maximal rectangle for each row
            maxRectangle = Solution.calculateMaxRectangle(heights, leftBoundaries, rightBoundaries, maxRectangle);
        }

        return maxRectangle;
    }

    /**
     * Updates the heights and left boundaries of rectangles for a given row.
     *
     * @param row            The row in the matrix.
     * @param heights        The array storing the heights of rectangles.
     * @param leftBoundaries The array storing the left boundaries of rectangles.
     * @param left           The left boundary for the current row.
     */
    private static void updateHeightsAndLeftBoundaries(
            final char[] row, final int[] heights, final int[] leftBoundaries, int left) {
        int i = left;
        for (int j = 0; j < heights.length; j++) {
            if ((int) '1' == (int) row[j]) {
                heights[j]++;
                leftBoundaries[j] = Math.max(leftBoundaries[j], i);
            } else {
                heights[j] = 0;
                leftBoundaries[j] = 0;
                i = j + 1;
            }
        }
    }

    /**
     * Updates the right boundaries of rectangles for a given row.
     *
     * @param row             The row in the matrix.
     * @param rightBoundaries The array storing the right boundaries of rectangles.
     * @param right           The right boundary for the current row.
     */
    private static void updateRightBoundaries(final char[] row, final int[] rightBoundaries, int right) {
        int i = right;
        for (int j = rightBoundaries.length - 1; j >= 0; j--) {
            if ((int) '1' == (int) row[j]) {
                rightBoundaries[j] = Math.min(rightBoundaries[j], i);
            } else {
                rightBoundaries[j] = i;
                i = j;
            }
        }
    }

    /**
     * Calculates the maximal rectangle for a given row.
     *
     * @param heights         The array storing the heights of rectangles.
     * @param leftBoundaries  The array storing the left boundaries of rectangles.
     * @param rightBoundaries The array storing the right boundaries of rectangles.
     * @param maxRectangle    The maximal rectangle found so far.
     *
     * @return The maximal rectangle found after considering the current row.
     */
    private static int calculateMaxRectangle(
            final int[] heights, final int[] leftBoundaries, final int[] rightBoundaries, int maxRectangle) {
        int rectangle = maxRectangle;
        for (int j = 0; j < heights.length; j++) {
            final int width = rightBoundaries[j] - leftBoundaries[j];
            final int area = heights[j] * width;
            rectangle = Math.max(rectangle, area);
        }
        return rectangle;
    }
}
